netaddr.go

```diff
@@ -220,10 +200,152 @@ func parseIPv4(s string) (ip IP, ok bool) {
 		s = s[j:]
 	}
 	if len(s) != 0 {
-		return IP{}, false
+		if s[0] == '%' {
+			return IP{}, parseErr(s, "zone specifier not allowed with IPv4 addresses")
+		}
+		return IP{}, parseErr(s, "trailing garbage after address")
+	}
+
+	return IPv4(ip4[0], ip4[1], ip4[2], ip4[3]), nil
+}
+
+// parseIPv6 parses s as an IPv6 address (in form "2001:db8::68").
+func parseIPv6(s string) (IP, error) {
+	// Split off the zone right from the start. Yes it's a second scan
+	// of the string, but trying to handle it inline makes a bunch of
+	// other inner loop conditionals more expensive, and it ends up
+	// being slower.
+	zone := ""
+	i := strings.IndexByte(s, '%')
+	if i != -1 {
+		s, zone = s[:i], s[i+1:]
+		if zone == "" {
+			// Not allowed to have an empty zone if explicitly specified.
+			return IP{}, parseErr(s, "zone must be a non-empty string")
+		}
 	}
 
-	return IPv4(ip4[0], ip4[1], ip4[2], ip4[3]), true
+	var ip [16]byte
+	ellipsis := -1 // position of ellipsis in ip
+
+	// Might have leading ellipsis
+	if len(s) >= 2 && s[0] == ':' && s[1] == ':' {
+		ellipsis = 0
+		s = s[2:]
+		// Might be only ellipsis
+		if len(s) == 0 {
+			return IPv6Unspecified().WithZone(zone), nil
+		}
+	}
+
+	// Loop, parsing hex numbers followed by colon.
+	i = 0
+	for i < 16 {
+		// Hex number. Similar to parseIPv4, inlining the hex number
+		// parsing yields a significant performance increase.
+		off := 0
+		acc := uint32(0)
+		for ; off < len(s); off++ {
+			c := s[off]
+			if c >= '0' && c <= '9' {
+				acc = (acc << 4) + uint32(c-'0')
+			} else if c >= 'a' && c <= 'f' {
+				acc = (acc << 4) + uint32(c-'a'+10)
+			} else if c >= 'A' && c <= 'F' {
+				acc = (acc << 4) + uint32(c-'A'+10)
+			} else {
+				break
+			}
+			if acc > maxUint16 {
+				// Overflow, fail.
+				return IP{}, parseErr(s, "IPv6 field has value >2^16")
```

Comment:
`>=`?

Reply:
I think `>` would be correct because `ffff` would be 2^16 and valid, yeah?

Reply:
```python
>>> 0xffff
65535
>>> 2**16
65536
```

Reply:
maxUint16 is 2^16-1, so > is correct. >= would reject the input `::ffff`, which is a valid IP.

Reply:
Sorry to be unclear. I am referring to the error message string, not the comparison two lines up.

Reply:
oh, doh. Yes.